dvc/output/__init__.py

```diff
@@ -1,136 +1,390 @@
+                        self
+                    )
+                )
+            return
+
+        assert not self.dependency
+
+        if not self.changed():
+            logger.info(
+                "Output '{}' didn't change. Skipping saving.".format(self)
             )
+            return
+
+        if self.is_in_repo:
+            if self.repo.scm.is_tracked(self.path):
+                raise OutputAlreadyTrackedError(self)
+
+            if self.use_cache:
+                self.repo.scm.ignore(self.path)
+
+        self.info = self.remote.save_info(self.path_info)
+
+    def commit(self):
+        if self.use_cache:
+            self.cache.save(self.path_info, self.info)
+
+    def dumpd(self):
+        ret = copy(self.info)
+        ret[self.PARAM_PATH] = self.url
+
+        if self.dependency:
+            return ret
+
+        ret[self.PARAM_CACHE] = self.use_cache
+
+        if isinstance(self.metric, dict):
+            if (
+                self.PARAM_METRIC_XPATH in self.metric
+                and not self.metric[self.PARAM_METRIC_XPATH]
+            ):
+                del self.metric[self.PARAM_METRIC_XPATH]
+
+        ret[self.PARAM_METRIC] = self.metric
+        ret[self.PARAM_PERSIST] = self.persist
+
+        if self.tags:
+            ret[self.PARAM_TAGS] = self.tags
+
+        return ret
+
+    def verify_metric(self):
+        raise DvcException(
+            "verify metric is not supported for {}".format(self.scheme)
+        )
+
+    def download(self, to_info, resume=False):
+        self.remote.download([self.path_info], [to_info], resume=resume)
+
+    def checkout(self, force=False, progress_callback=None, tag=None):
+        if not self.use_cache:
+            return
+
+        if tag:
+            info = self.tags[tag]
+        else:
+            info = self.info
+
+        self.cache.checkout(
+            self.path_info,
+            info,
+            force=force,
+            progress_callback=progress_callback,
         )
-    return ret
+
+    def remove(self, ignore_remove=False):
+        self.remote.remove(self.path_info)
+        if self.scheme != "local":
+            return
+
+        if ignore_remove and self.use_cache and self.is_in_repo:
+            self.repo.scm.ignore_remove(self.path)
+
+    def move(self, out):
+        if self.scheme == "local" and self.use_cache and self.is_in_repo:
+            self.repo.scm.ignore_remove(self.path)
+
+        self.remote.move(self.path_info, out.path_info)
+        self.url = out.url
+        self.path_info = copy(out.path_info)
+        self.save()
+        self.commit()
+
+        if self.scheme == "local" and self.use_cache and self.is_in_repo:
+            self.repo.scm.ignore(self.path)
+
+    def get_files_number(self):
+        if not self.use_cache or not self.checksum:
+            return 0
+
+        if self.is_dir_checksum:
+            return len(self.dir_cache)
+
+        return 1
+
+    def unprotect(self):
```

Comment:
does it make sense for a dependency? in general - do all the methods make sense for both output and dependency?

Reply:
Agreed, after refactor this method only exists in Output.